File: modules/native-alert/NativeAlertModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

import { type NativeAlertOptions } from './native-alert.types';

// ─── requireNativeModule vs requireOptionalNativeModule ───────────────────────
//
// requireNativeModule('NativeAlert')
//   → 找不到模組時【立即拋出例外】並讓 App 崩潰
//   → 適合：確定已編譯進原生 App、絕對不能缺少的核心模組
//
// requireOptionalNativeModule('NativeAlert')
//   → 找不到模組時【回傳 null】，不崩潰
//   → 適合：示範用、漸進增強、Expo Go 相容、或有 JS fallback 的場景
//
// 本專案選用 requireOptionalNativeModule 的理由：
//   1. 在 Expo Go 環境中，自訂原生模組不存在，但 App 不應崩潰
//   2. 在尚未執行 `expo prebuild` / `yarn ios / android` 時，同樣不應崩潰
//   3. index.native.ts 已提供原生呼叫，index.ts 提供 JS fallback
//      → 這層只負責「橋接」，需要能安全地返回 null 讓上層決定如何處理
//
// 若要強制要求原生模組存在（生產環境核心功能），請改回 requireNativeModule
// ──────────────────────────────────────────────────────────────────────────────

type NativeAlertNativeModuleType = NativeModule & {
  showAlert: (
    title: string,
    message: string | undefined,
    buttons: { text: string; style?: string }[] | undefined
  ) => Promise<string | null>;
};

// 找不到原生模組時回傳 null（不崩潰），由 index.native.ts 的呼叫端處理 null 情境
const NativeAlertNativeModule =
  requireOptionalNativeModule<NativeAlertNativeModuleType>('NativeAlert');

export { NativeAlertNativeModule };
export type { NativeAlertOptions };

File: modules/native-alert/android/src/main/java/expo/modules/nativealert/NativeAlertModule.kt
package expo.modules.nativealert

import android.app.AlertDialog
import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

// MARK: - NativeAlertModule

/**
 * 原生 Alert 模組（Android 實作）
 * 使用 AlertDialog 呈現系統原生 Alert
 *
 * 注意：這個 local module 不使用 expo-module-gradle-plugin，
 * 所以無法使用 appContext.mainQueue.launch{}（需要 kotlinx.coroutines）。
 * 改用 Handler(Looper.getMainLooper()).post{} 達到相同的 Main Thread 切換效果。
 */
class NativeAlertModule : Module() {

  // Module 定義
  override fun definition() = ModuleDefinition {
    // 模組名稱：JS 端用這個名字找到此模組
    Name("NativeAlert")

    // 定義 JS 可呼叫的 async function
    AsyncFunction("showAlert") { title: String, message: String?, buttons: List<Map<String, String>>?, promise: Promise ->
      val activity = appContext.currentActivity
      if (activity === null) {
        promise.reject("NO_ACTIVITY", "找不到目前的 Activity", null)
        return@AsyncFunction
      }

      // 所有 UI 操作必須在 Main Thread 執行
      // 使用 Handler + Looper 取代 coroutines（不需要額外依賴）
      Handler(Looper.getMainLooper()).post {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(title)
        if (!message.isNullOrEmpty()) {
          builder.setMessage(message)
        }

        // 如果有傳入按鈕設定，就逐一設定
        if (!buttons.isNullOrEmpty()) {
          for (buttonConfig in buttons) {
            val buttonText = buttonConfig["text"] ?: "OK"
            val buttonStyle = buttonConfig["style"] ?: "default"

            when (buttonStyle) {
              "cancel" -> builder.setNegativeButton(buttonText) { _, _ ->
                promise.resolve(buttonText)
              }
              "destructive" -> builder.setNeutralButton(buttonText) { _, _ ->
                promise.resolve(buttonText)
              }
              else -> builder.setPositiveButton(buttonText) { _, _ ->
                promise.resolve(buttonText)
              }
            }
          }
        } else {
          // 沒有設定按鈕時，預設加一個 OK
          builder.setPositiveButton("OK") { _, _ ->
            promise.resolve("OK")
          }
        }

        // 取消時也要 resolve，避免 Promise 永遠 pending
        builder.setOnCancelListener {
          promise.resolve(null)
        }

        builder.show()
      }
    }
  }
}
